common/beatmap_utils.py

```diff
@@ -0,0 +1,472 @@
+        for beatmap in beatmaps:
+            await save(beatmap)
+
+    for beatmap in beatmaps:
+        if beatmap.id == id:
+            return beatmap
+
+
+async def set_from_api(id: int, should_save: bool = True) -> Optional[list[Beatmap]]:
+    api_key = random.choice(settings.OSU_API_KEYS)
+
+    response = await glob.http_client.get(
+        GET_BEATMAP_URL,
+        params={"k": api_key, "s": id},
+    )
+    if response.status_code == 404:
+        return None
+
+    response.raise_for_status()
+
+    response_json = response.json()
+    if not response_json:
+        return None
+
+    beatmaps = parse_from_osu_api(response_json)
+
+    if should_save:
+        for beatmap in beatmaps:
+            await save(beatmap)
+
+    return beatmaps
+
+
+IGNORED_BEATMAP_CHARS = dict.fromkeys(map(ord, r':\/*<>?"|'), None)
+
+FROZEN_STATUSES = (RankedStatus.RANKED, RankedStatus.APPROVED, RankedStatus.LOVED)
+
+
+def parse_from_osu_api(
+    response_json_list: list[dict],
+    frozen: bool = False,
+) -> list[Beatmap]:
+    maps = []
+
+    for response_json in response_json_list:
+        md5 = response_json["file_md5"]
+        id = int(response_json["beatmap_id"])
+        set_id = int(response_json["beatmapset_id"])
+
+        filename = (
+            ("{artist} - {title} ({creator}) [{version}].osu")
+            .format(**response_json)
+            .translate(IGNORED_BEATMAP_CHARS)
+        )
+
+        song_name = (
+            ("{artist} - {title} [{version}]")
+            .format(**response_json)
+            .translate(IGNORED_BEATMAP_CHARS)
+        )
+
+        hit_length = int(response_json["hit_length"])
+
+        if _max_combo := response_json.get("max_combo"):
+            max_combo = int(_max_combo)
+        else:
+            max_combo = 0
+
+        ranked_status = RankedStatus.from_osu_api(int(response_json["approved"]))
+        if ranked_status in FROZEN_STATUSES:
+            frozen = True  # beatmaps are always frozen when ranked/approved/loved
+
+        mode = int(response_json["mode"])
+
+        if _bpm := response_json.get("bpm"):
+            bpm = round(float(_bpm))
+        else:
+            bpm = 0
+
+        od = float(response_json["diff_overall"])
+        ar = float(response_json["diff_approach"])
+
+        maps.append(
+            Beatmap(
+                md5=md5,
+                id=id,
+                set_id=set_id,
+                song_name=song_name,
+                status=ranked_status,
+                plays=0,
+                passes=0,
+                mode=mode,
+                od=od,
+                ar=ar,
+                hit_length=hit_length,
+                last_update=int(time.time()),
+                max_combo=max_combo,
+                bpm=bpm,
+                filename=filename,
+                frozen=frozen,
+                rating=10.0,
+            ),
+        )
+
+    return maps
+
+
+async def increment_playcount(beatmap: Beatmap, passcount: bool = True) -> None:
```

Comment:
i would just get rid of this since its unused in bancho (also missing flame's atomicity fix)